### PR TITLE
add m1 requestability - WIP

### PR DESCRIPTION
**Pull request by charmingduchess**
A lot went down:
- removed requestability logic from AvailabilityResolver
- refactored deliveryLocationsResolver into more reusable functions 
- created RequestabilityResolver, which calls addRecap or addOnsite locations functions from deliveryLocationsResolver
   - earlier version of this code had an extra scsb call which i realized is not necessary because the recap customer codes are added in the availabilityresolver call. i still think there should be some further centralization of the scsb calls and there should not be batched scsb queries being invoked from functions defined within the resolvers 
- removed fixPhys and fixEddRequestability functions. phys requestablilty is determined from a non zero length of delivery locations and eddRequestability is determined using functions from the deliveryLocationsResolver
- removed support for no-onsite-edd request options

The code has LOTS of unit tests, many of which I've moved around as I've moved parts of the code around. A bunch of the tests in `requestability_resolver.test.js` came from availabilityResolver, for example. I think even before looking at the code, it would be most useful to go through the test cases and check that the output is in line with our business logic. 

NB: To run the tests, run `NYPL_CORE_VERSION=romcom-2.0-prerelease npm test`

Not in this PR:
-  consideration for batching
- consideration of what happens when the requestabilityResolver is invoked via the responseMassager.

### CODE REVIEW

File: lib/requestability_resolver.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class RequestabilityResolver {
+  static fixItemRequestability (elasticSearchResponse) {
+    elasticSearchResponse.hits.hits
+      .map((hit) => {
+        hit._source.items = hit._source.items.map(function (item) {
+          if (item.electronicLocator) return item
+          if (item.recapCustomerCode && item.recapCustomerCode[0]) {
+            item = DeliveryLocationsResolver.attachRecapDeliveryInfo(item, item.recapCustomerCode[0])
+          } else {
+            item = DeliveryLocationsResolver.attachOnsiteDeliveryInfo(item, item.recapCustomerCode)
```
Comment (nonword) [inline lib/requestability_resolver.js:11]: This attaches `.deliveryLocation` to all items including all items served through search results, which could be significant extra bytes to download for some search results (and a violation of our swagger), so ideally we only serve delivery locations through the deliveryLocationsByBarcode endpoint.

Could `DeliveryLocationsResolver` export something like a `.getOnsiteDeliveryInfo(item)` that returns the same properties? (While still exporting these attach functions, which would just be convenience methods to make internal calls to `getOnsiteDeliveryInfo` etc and attach the props to the items?)

File: lib/requestability_resolver.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+const DeliveryLocationsResolver = require('./delivery-locations-resolver')
+class RequestabilityResolver {
+  static fixItemRequestability (elasticSearchResponse) {
+    elasticSearchResponse.hits.hits
+      .map((hit) => {
+        hit._source.items = hit._source.items.map(function (item) {
+          if (item.electronicLocator) return item
+          if (item.recapCustomerCode && item.recapCustomerCode[0]) {
+            item = DeliveryLocationsResolver.attachRecapDeliveryInfo(item, item.recapCustomerCode[0])
```
Comment (nonword) [inline lib/requestability_resolver.js:9]: Given the name of this call, why pass in `item.recapCustomerCode` as second param when the resolver can get it from the first param?

File: lib/requestability_resolver.js (hunk 1 above)
Comment (nonword) [reply lib/requestability_resolver.js:11]: Also, no second param allowed in `attachOnsiteDeliveryInfo`

**Review by nonword — COMMENTED** [review]
Looking good. See comments

File: lib/delivery-locations-resolver.js [hunk 3]
```diff
@@ -160,24 +167,118 @@ class DeliveryLocationsResolver {
...
+      // for a made up location code.
+      logger.warn('Malformed or missing holding location code: ', holdingLocationCode)
+      deliveryInfo.eddRequestable = false
+      deliveryInfo.deliveryLocation = []
+      return deliveryInfo
+    }
+    deliveryInfo.eddRequestable = this.eddRequestableByOnSiteCriteria(item)
+    // unrequestable holding location overrides deliveryLocations potentially in nypl-core
+    if (!this.requestableBasedOnHoldingLocation(item)) {
+      return deliveryInfo
```
Comment (nonword) [inline lib/delivery-locations-resolver.js:257]: If holding location is not requestable return an empty `{}`?

File: lib/delivery-locations-resolver.js [hunk 4]
```diff
@@ -160,24 +167,118 @@ class DeliveryLocationsResolver {
...
+  static getOnsiteDeliveryInfo (item) {
+    let deliveryInfo = {}
+    const holdingLocationCode = this.extractLocationCode(item)
+    const sierraData = sierraLocations[holdingLocationCode]
+    if (!sierraData) {
+      // This case is mainly to satisfy a test which wants eddRequestable = false
+      // for a made up location code.
+      logger.warn('Malformed or missing holding location code: ', holdingLocationCode)
+      deliveryInfo.eddRequestable = false
+      deliveryInfo.deliveryLocation = []
```
Comment (nonword) [inline lib/delivery-locations-resolver.js:251]: May be better to initialize these defaults up at line 243? (i.e. default is eddRequestable false and empty deliveryLocations)

File: lib/delivery-locations-resolver.js [hunk 5]
```diff
@@ -160,24 +167,118 @@ class DeliveryLocationsResolver {
...
+    }
+    deliveryInfo.eddRequestable = this.eddRequestableByOnSiteCriteria(item)
+    // unrequestable holding location overrides deliveryLocations potentially in nypl-core
+    if (!this.requestableBasedOnHoldingLocation(item)) {
+      return deliveryInfo
+    }
+    // if nypl-core reports that a holding location's delivery locations
+    // should be found by M2 code
+    const deliverableToResolution = sierraData.deliverableToResolution
+    if (deliverableToResolution === 'm2-customer-codes') {
```
Comment (nonword) [inline lib/delivery-locations-resolver.js:262]: Seems like it should be 'm2-customer-code'?

**Comment by nonword** [comment]
Also, does this work depend on @nypl/nypl-core-objects v1.6.1?

**Review by nonword — COMMENTED** [review]
Looking good. See comments

File: lib/delivery-locations-resolver.js [hunk 6]
```diff
@@ -160,85 +167,144 @@ class DeliveryLocationsResolver {
...
+    // but its definitely not phys requestable.
+    deliveryInfo.eddRequestable = this.eddRequestableByOnSiteCriteria(item)
+    if (!this.requestableBasedOnHoldingLocation(item)) {
+      return deliveryInfo
+    }
+    // if nypl-core reports that a holding location's delivery locations
+    // should be found by M2 code
+    const deliverableToResolution = sierraData.deliverableToResolution
+    if (deliverableToResolution === 'm2-customer-code') {
+      deliveryInfo.deliveryLocation = this.deliveryLocationsByM2CustomerCode(item.m2CustomerCode[0])
```
[the remark this answers is not in the document]
Comment (charmingduchess) [reply lib/delivery-locations-resolver.js:265]: so if there's no m2 customer code, is physRequestable false?

[the remark this answers is not in the document]
Comment (nonword) [reply lib/delivery-locations-resolver.js:265]: Yes. If the item is in M2 and there's no M2 customer code, we can't possibly determine delivery options, so we should not mark it physRequestable

**Review by nonword — APPROVED** [review]
Looks good

**Pull request merged**